File: Sources/SSHTunnelManager/Models/TextDiff.swift
import Foundation

/// Per-row diff status shared with the Scintilla bridge. The raw `Int` values
/// MUST stay in sync with the bridge's interpretation in `SciEditorView.mm`.
enum DiffRowStatus: Int {
    case equal = 0     // identical on both sides
    case added = 1     // present only on the right
    case deleted = 2   // present only on the left
    case changed = 3   // present on both sides but different
    case filler = 4    // a blank spacer inserted to keep the two sides aligned
}

/// One side (left or right) of an aligned diff, already padded with filler rows
/// so both sides have exactly the same number of rows.
struct DiffSide {
    /// Rows joined with "\n"; filler rows are empty lines.
    var text: String
    /// `DiffRowStatus.rawValue` for each row.
    var status: [Int]
    /// Original 1-based line number for each row (0 for filler rows).
    var numbers: [Int]
    /// Start of an intra-line change, in UTF-8 bytes within the row (0 = none).
    var spanStart: [Int]
    /// Length of the intra-line change, in UTF-8 bytes (0 = no intra-line span).
    var spanLength: [Int]
}

/// The result of aligning two documents line-by-line.
struct DiffResult {
    var left: DiffSide
    var right: DiffSide
}

/// A small, dependency-free line differ used by the Scintilla compare view.
///
/// It interns lines to integers, trims the common prefix/suffix, runs a classic
/// longest-common-subsequence alignment on the middle, then pairs adjacent
/// deletes/inserts into "changed" rows and pads each side with filler rows so
/// the two panes line up. For changed rows it also computes a single intra-line
/// highlight span (the differing middle after trimming a common prefix/suffix).
enum TextDiff {

    /// The three primitive alignment operations, carrying original indices.
    private enum Op {
        case equal(Int, Int)   // (leftIndex, rightIndex)
        case del(Int)          // leftIndex only
        case ins(Int)          // rightIndex only
    }

    static func compare(_ leftText: String, _ rightText: String) -> DiffResult {
        let a = leftText.components(separatedBy: "\n")
        let b = rightText.components(separatedBy: "\n")

        // Intern lines to integers so the alignment compares Ints, not Strings.
        var table: [String: Int] = [:]
        table.reserveCapacity(a.count + b.count)
        func intern(_ s: String) -> Int {
            if let v = table[s] { return v }
            let v = table.count; table[s] = v; return v
        }
        let ai = a.map(intern)
        let bi = b.map(intern)

        let ops = diffOps(ai, bi)

        var left = SideBuilder()
        var right = SideBuilder()
        var pendingDel: [Int] = []
        var pendingIns: [Int] = []

        func flushGap() {
            if pendingDel.isEmpty && pendingIns.isEmpty { return }
            let k = min(pendingDel.count, pendingIns.count)
            // Pair up as many delete/insert lines as possible into "changed".
            for i in 0..<k {
                let li = pendingDel[i], ri = pendingIns[i]
                let (ls, ll, rs, rl) = intralineSpan(a[li], b[ri])
                left.add(a[li], .changed, li + 1, ls, ll)
                right.add(b[ri], .changed, ri + 1, rs, rl)
            }
            // Remaining deletes: real on the left, filler on the right.
            for i in k..<pendingDel.count {
                let li = pendingDel[i]
                left.add(a[li], .deleted, li + 1)
                right.add("", .filler, 0)
            }
            // Remaining inserts: filler on the left, real on the right.
            for i in k..<pendingIns.count {
                let ri = pendingIns[i]
                left.add("", .filler, 0)
                right.add(b[ri], .added, ri + 1)
            }
            pendingDel.removeAll(keepingCapacity: true)
            pendingIns.removeAll(keepingCapacity: true)
        }

        for op in ops {
            switch op {
            case .equal(let li, let ri):
                flushGap()
                left.add(a[li], .equal, li + 1)
                right.add(b[ri], .equal, ri + 1)
            case .del(let li):
                pendingDel.append(li)
            case .ins(let ri):
                pendingIns.append(ri)
            }
        }
        flushGap()

        return DiffResult(left: left.side(), right: right.side())
    }

    // MARK: - Alignment

    /// Produces an ordered list of equal/delete/insert operations, trimming the
    /// common prefix and suffix first so the LCS only runs on the middle.
    private static func diffOps(_ a: [Int], _ b: [Int]) -> [Op] {
        var ops: [Op] = []

        var lo = 0
        let n = a.count, m = b.count
        while lo < n && lo < m && a[lo] == b[lo] {
            ops.append(.equal(lo, lo)); lo += 1
        }

        var hiA = n, hiB = m
        var suffix: [Op] = []
        while hiA > lo && hiB > lo && a[hiA - 1] == b[hiB - 1] {
            hiA -= 1; hiB -= 1
            suffix.append(.equal(hiA, hiB))
        }
        suffix.reverse()

        ops.append(contentsOf: lcsOps(a, b, lo: lo, hiA: hiA, hiB: hiB))
        ops.append(contentsOf: suffix)
        return ops
    }

    /// Classic LCS over `a[lo..<hiA]` vs `b[lo..<hiB]`, emitting ops with the
    /// original indices. Falls back to a full replace for very large middles.
    private static func lcsOps(_ a: [Int], _ b: [Int], lo: Int, hiA: Int, hiB: Int) -> [Op] {
        let n = hiA - lo, m = hiB - lo
        if n == 0 && m == 0 { return [] }
        if n == 0 { return (lo..<hiB).map { .ins($0) } }
        if m == 0 { return (lo..<hiA).map { .del($0) } }
        if n * m > 4_000_000 {
            // Too large to align precisely; show the middle as a full replace.
            var ops: [Op] = (lo..<hiA).map { .del($0) }
            ops.append(contentsOf: (lo..<hiB).map { .ins($0) })
            return ops
        }

        // dp[i][j] = LCS length of a[lo+i ..< hiA] and b[lo+j ..< hiB].
        let cols = m + 1
        var dp = [Int](repeating: 0, count: (n + 1) * cols)
        for i in stride(from: n - 1, through: 0, by: -1) {
            for j in stride(from: m - 1, through: 0, by: -1) {
                if a[lo + i] == b[lo + j] {
                    dp[i * cols + j] = dp[(i + 1) * cols + (j + 1)] + 1
                } else {
                    dp[i * cols + j] = max(dp[(i + 1) * cols + j], dp[i * cols + (j + 1)])
                }
            }
        }

        var ops: [Op] = []
        var i = 0, j = 0
        while i < n && j < m {
            if a[lo + i] == b[lo + j] {
                ops.append(.equal(lo + i, lo + j)); i += 1; j += 1
            } else if dp[(i + 1) * cols + j] >= dp[i * cols + (j + 1)] {
                ops.append(.del(lo + i)); i += 1
            } else {
                ops.append(.ins(lo + j)); j += 1
            }
        }
        while i < n { ops.append(.del(lo + i)); i += 1 }
        while j < m { ops.append(.ins(lo + j)); j += 1 }
        return ops
    }

    // MARK: - Intra-line span

    /// Trims a common prefix and suffix (by character) from two differing lines
    /// and returns the byte offset + byte length of the changed middle on each
    /// side, ready to feed Scintilla's UTF-8 positions.
    private static func intralineSpan(_ x: String, _ y: String)
        -> (leftStart: Int, leftLen: Int, rightStart: Int, rightLen: Int) {
        let xc = Array(x), yc = Array(y)
        var p = 0
        while p < xc.count && p < yc.count && xc[p] == yc[p] { p += 1 }
        var s = 0
        while s < (xc.count - p) && s < (yc.count - p)
              && xc[xc.count - 1 - s] == yc[yc.count - 1 - s] { s += 1 }

        let leftPrefix = String(xc[0..<p])
        let leftMiddle = String(xc[p..<(xc.count - s)])
        let rightPrefix = String(yc[0..<p])
        let rightMiddle = String(yc[p..<(yc.count - s)])

        return (leftPrefix.utf8.count, leftMiddle.utf8.count,
                rightPrefix.utf8.count, rightMiddle.utf8.count)
    }

    // MARK: - Side accumulator

    private struct SideBuilder {
        private var lines: [String] = []
        private var status: [Int] = []
        private var numbers: [Int] = []
        private var spanStart: [Int] = []
        private var spanLength: [Int] = []

        mutating func add(_ line: String, _ st: DiffRowStatus, _ number: Int,
                          _ start: Int = 0, _ length: Int = 0) {
            lines.append(line)
            status.append(st.rawValue)
            numbers.append(number)
            spanStart.append(start)
            spanLength.append(length)
        }

        func side() -> DiffSide {
            DiffSide(text: lines.joined(separator: "\n"),
                     status: status, numbers: numbers,
                     spanStart: spanStart, spanLength: spanLength)
        }
    }
}

File: cross-platform/src/RemoteStuff/Services/TextDiff.cs
using System;
using System.Collections.Generic;

namespace RemoteStuff.Services;

/// <summary>Per-row diff classification for the two-pane compare view.</summary>
public enum DiffRowStatus
{
    Equal = 0,   // identical on both sides
    Added = 1,   // present only on the right
    Deleted = 2, // present only on the left
    Changed = 3, // present on both sides but different
    Filler = 4,  // a blank spacer inserted to keep the two sides aligned
}

/// <summary>One aligned row of a side-by-side diff. Both sides always share the
/// same number of rows; filler rows keep them lined up.</summary>
public sealed class DiffRow
{
    public int LeftNumber { get; init; }        // 1-based; 0 for filler
    public string LeftText { get; init; } = "";
    public DiffRowStatus LeftStatus { get; init; }

    public int RightNumber { get; init; }       // 1-based; 0 for filler
    public string RightText { get; init; } = "";
    public DiffRowStatus RightStatus { get; init; }

    public string LeftNumberText => LeftNumber > 0 ? LeftNumber.ToString() : "";
    public string RightNumberText => RightNumber > 0 ? RightNumber.ToString() : "";
}

/// <summary>A small, dependency-free line differ (a C# port of the macOS app's
/// <c>TextDiff</c>). Interns lines to integers, trims the common prefix/suffix,
/// runs a classic LCS over the middle, then pairs adjacent deletes/inserts into
/// "changed" rows and pads each side with filler rows so the panes line up.</summary>
public static class TextDiff
{
    private enum Kind { Equal, Del, Ins }

    private readonly struct Op
    {
        public readonly Kind Kind;
        public readonly int Left;
        public readonly int Right;
        public Op(Kind kind, int left, int right) { Kind = kind; Left = left; Right = right; }
    }

    public static List<DiffRow> Compare(string leftText, string rightText)
    {
        var a = (leftText ?? "").Replace("\r\n", "\n").Replace("\r", "\n").Split('\n');
        var b = (rightText ?? "").Replace("\r\n", "\n").Replace("\r", "\n").Split('\n');

        // Intern lines to integers so the alignment compares ints, not strings.
        var table = new Dictionary<string, int>(a.Length + b.Length);
        int Intern(string s)
        {
            if (table.TryGetValue(s, out var v)) return v;
            v = table.Count; table[s] = v; return v;
        }
        var ai = new int[a.Length];
        var bi = new int[b.Length];
        for (int i = 0; i < a.Length; i++) ai[i] = Intern(a[i]);
        for (int i = 0; i < b.Length; i++) bi[i] = Intern(b[i]);

        var ops = DiffOps(ai, bi);

        var rows = new List<DiffRow>();
        var pendingDel = new List<int>();
        var pendingIns = new List<int>();

        void FlushGap()
        {
            if (pendingDel.Count == 0 && pendingIns.Count == 0) return;
            int k = Math.Min(pendingDel.Count, pendingIns.Count);
            for (int i = 0; i < k; i++)
            {
                int li = pendingDel[i], ri = pendingIns[i];
                rows.Add(new DiffRow
                {
                    LeftNumber = li + 1, LeftText = a[li], LeftStatus = DiffRowStatus.Changed,
                    RightNumber = ri + 1, RightText = b[ri], RightStatus = DiffRowStatus.Changed,
                });
            }
            for (int i = k; i < pendingDel.Count; i++)
            {
                int li = pendingDel[i];
                rows.Add(new DiffRow
                {
                    LeftNumber = li + 1, LeftText = a[li], LeftStatus = DiffRowStatus.Deleted,
                    RightNumber = 0, RightText = "", RightStatus = DiffRowStatus.Filler,
                });
            }
            for (int i = k; i < pendingIns.Count; i++)
            {
                int ri = pendingIns[i];
                rows.Add(new DiffRow
                {
                    LeftNumber = 0, LeftText = "", LeftStatus = DiffRowStatus.Filler,
                    RightNumber = ri + 1, RightText = b[ri], RightStatus = DiffRowStatus.Added,
                });
            }
            pendingDel.Clear();
            pendingIns.Clear();
        }

        foreach (var op in ops)
        {
            switch (op.Kind)
            {
                case Kind.Equal:
                    FlushGap();
                    rows.Add(new DiffRow
                    {
                        LeftNumber = op.Left + 1, LeftText = a[op.Left], LeftStatus = DiffRowStatus.Equal,
                        RightNumber = op.Right + 1, RightText = b[op.Right], RightStatus = DiffRowStatus.Equal,
                    });
                    break;
                case Kind.Del:
                    pendingDel.Add(op.Left);
                    break;
                case Kind.Ins:
                    pendingIns.Add(op.Right);
                    break;
            }
        }
        FlushGap();
        return rows;
    }

    /// <summary>True when the two texts differ line-for-line.</summary>
    public static bool AreDifferent(string leftText, string rightText)
    {
        var l = (leftText ?? "").Replace("\r\n", "\n").Replace("\r", "\n");
        var r = (rightText ?? "").Replace("\r\n", "\n").Replace("\r", "\n");
        return !string.Equals(l, r, StringComparison.Ordinal);
    }

    private static List<Op> DiffOps(int[] a, int[] b)
    {
        var ops = new List<Op>();
        int lo = 0, n = a.Length, m = b.Length;
        while (lo < n && lo < m && a[lo] == b[lo])
        {
            ops.Add(new Op(Kind.Equal, lo, lo)); lo++;
        }

        int hiA = n, hiB = m;
        var suffix = new List<Op>();
        while (hiA > lo && hiB > lo && a[hiA - 1] == b[hiB - 1])
        {
            hiA--; hiB--;
            suffix.Add(new Op(Kind.Equal, hiA, hiB));
        }
        suffix.Reverse();

        ops.AddRange(LcsOps(a, b, lo, hiA, hiB));
        ops.AddRange(suffix);
        return ops;
    }

    private static List<Op> LcsOps(int[] a, int[] b, int lo, int hiA, int hiB)
    {
        int n = hiA - lo, m = hiB - lo;
        var ops = new List<Op>();
        if (n == 0 && m == 0) return ops;
        if (n == 0)
        {
            for (int j = lo; j < hiB; j++) ops.Add(new Op(Kind.Ins, 0, j));
            return ops;
        }
        if (m == 0)
        {
            for (int i = lo; i < hiA; i++) ops.Add(new Op(Kind.Del, i, 0));
            return ops;
        }
        if ((long)n * m > 4_000_000)
        {
            for (int i = lo; i < hiA; i++) ops.Add(new Op(Kind.Del, i, 0));
            for (int j = lo; j < hiB; j++) ops.Add(new Op(Kind.Ins, 0, j));
            return ops;
        }

        int cols = m + 1;
        var dp = new int[(n + 1) * cols];
        for (int i = n - 1; i >= 0; i--)
        {
            for (int j = m - 1; j >= 0; j--)
            {
                if (a[lo + i] == b[lo + j])
                    dp[i * cols + j] = dp[(i + 1) * cols + (j + 1)] + 1;
                else
                    dp[i * cols + j] = Math.Max(dp[(i + 1) * cols + j], dp[i * cols + (j + 1)]);
            }
        }

        int x = 0, y = 0;
        while (x < n && y < m)
        {
            if (a[lo + x] == b[lo + y]) { ops.Add(new Op(Kind.Equal, lo + x, lo + y)); x++; y++; }
            else if (dp[(x + 1) * cols + y] >= dp[x * cols + (y + 1)]) { ops.Add(new Op(Kind.Del, lo + x, 0)); x++; }
            else { ops.Add(new Op(Kind.Ins, 0, lo + y)); y++; }
        }
        while (x < n) { ops.Add(new Op(Kind.Del, lo + x, 0)); x++; }
        while (y < m) { ops.Add(new Op(Kind.Ins, 0, lo + y)); y++; }
        return ops;
    }
}
